engineering-plans/0002-ethereum-tracing-cache.md

```diff
@@ -0,0 +1,201 @@
+
+We want to minimize the cost of external requests for trace data. It is likely that it is better to...
+ * Make fewer requests
+ * Not ask for trace data that is already cached
+ * Ask for trace data for multiple contract addresses within the same block when possible.
+
+ There is one flow of data which upholds these invariants. In doing so it makes a tradeoff of increasing latency for the execution of a specific subgraph, but increases throughput of the whole system.
+
+ Within this graph:
+ * Edges which are labelled refer to some subset of the output data.
+ * Edges which are not labelled refer to the entire set of the output data.
+ * Each node executes once for each contiguous range of blocks. That is, it merges all incoming data before executing, and executes the minimum possible times.
+ * The example given is just for 2 addresses. The actual code must work on sets of addresses.
+
+ ```mermaid
+ graph LR;
+    A[Block Range for Contract A & B]
+    A --> |Above Reorg Threshold| E
+    D[Get Cache A]
+    A --> |Below Reorg Threshold A| D
+    A --> |Below Reorg Threshold B| H
+    E[Ethereum A & B]
+    F[Ethereum A]
+    G[Ethereum B]
+    H[Get Cache B]
+    D --> |Found| M
+    H --> |Found| M
+    M[Result]
+    D --> |Missing| N
+    H --> |Missing| N
+    N[Overlap]
+    N --> |A & B| E
+    N --> |A| F
+    N --> |B| G
+    E --> M
+    K[Set Cache A]
+    L[Set Cache B]
+    E --> |B Below Reorg Threshold| L
+    E --> |A Below Reorg Threshold| K
+    F --> K
+    G --> L
+    F --> M
+    G --> M
+ ```
+
+ 
+This construction is designed to make the fewest number of the most efficient calls possible. It is not as complicated as it looks. The actual construction can be expressed as sequential steps with a set of filters preceding each step.
+
+### Useful dependencies
+The feature deals a lot with ranges and sets. Operations like sum, subtract, merge, and find overlapping are used frequently. [nested_intervals](https://crates.io/crates/nested_intervals) is a crate which provides some of these operations.
+
+
+## Tests
+
+### Benchmark
+A temporary benchmark will be added for indexing a simple subgraph which uses call handlers. The benchmark will be run in these scenarios:
+* Sync before changes
+* Re-sync before changes 
+* Sync after changes
+* Re-sync after changes
+
+### Ranges
+Due to the complexity of the resource minimizing data workflow, it will be useful to have mocks for the cache and database which record their calls, and check that expected calls are made for tricky data sets.
+
+### Database
+A real database integration test will be added to test the add/remove from cache implementation to verify that it correctly merges blocks, handles concurrency issues, etc.
+
+
+## Migration
+
+None
+
+## Documentation
+
+None, aside from code comments
+
+**Implementation Plan:**
```

Comment:
Let's make this `## Implementation Plan`, so it shows up as a proper section header in the document 😄

Reply:
Done.